tools/urtrace/collector.cpp

```diff
@@ -0,0 +1,303 @@
+    }
+
+    return ostr.str();
+}
+
+/*
+ * Since this is a library that gets loaded alongside the traced program, it
+ * can't just accept arguments from the trace CLI tool directly. Instead, the
+ * arguments are passed through an environment variable. Users shouldn't set
+ * these directly.
+ *
+ * Available arguments (documented in trace cli tool):
+ * - "print_begin"
+ * - "profiling"
+ * - "time_unit:<auto,ns, ...>"
+ * - "filter:<regex>"
+ */
+static class cli_args {
+    std::optional<std::string>
+    arg_with_value(std::string_view name, const EnvVarMap::value_type &arg) {
+        if (arg.first != name) {
+            return std::nullopt;
+        }
+        if (arg.second.empty() || arg.second.size() > 1) {
+            out.warn("{} requires a single argument, skipping...", name);
+            return std::nullopt;
+        }
+        return arg.second.at(0);
+    }
+
+  public:
+    cli_args() {
+        print_begin = false;
+        profiling = false;
+        time_unit = TIME_UNIT_AUTO;
+        if (auto args = getenv_to_map(ARGS_ENV, false)) {
+            for (auto arg : *args) {
+                if (arg.first == "print_begin") {
+                    print_begin = true;
+                } else if (arg.first == "profiling") {
+                    profiling = true;
+                } else if (arg.first == "no_args") {
+                    no_args = true;
+                } else if (auto unit = arg_with_value("time_unit", arg)) {
+                    for (int i = 0; i < MAX_TIME_UNIT; ++i) {
+                        if (time_unit_str[i] == unit) {
+                            time_unit = (enum time_unit)i;
+                        }
+                    }
+                } else if (auto filter_str = arg_with_value("filter", arg)) {
+                    try {
+                        filter = filter_str;
+                    } catch (const std::regex_error &err) {
+                        out.warn("invalid filter regex {} {}", *filter_str,
+                                 err.what());
+                    }
+                } else {
+                    out.warn("unknown {} argument {}.", ARGS_ENV, arg.first);
+                }
+            }
+        }
+        out.debug("collector args (.print_begin = {}, .profiling = {}, "
+                  ".time_unit = {}, .filter = {})",
+                  print_begin, profiling, time_unit_str[time_unit],
+                  filter_str.has_value() ? *filter_str : "none");
+    }
+
+    enum time_unit time_unit;
+    bool print_begin;
+    bool profiling;
+    bool no_args;
+    std::optional<std::string>
+        filter_str; //the filter_str is kept primarly for printing.
+    std::optional<std::regex> filter;
+} cli_args;
+
+using namespace ur_params;
+
+typedef std::chrono::high_resolution_clock Clock;
+
+class fn_context {
+  public:
+    std::chrono::time_point<Clock> start;
+};
+
+static thread_local std::map<uint64_t, fn_context> instance_data;
```

Comment:
why not unordered_map?

Reply:
I didn't give it much thought tbh. It's not performance critical. This map will likely contain only one element at a time, so a map should be better. What I wanted to use here is a stack, but I'm not sure if a function can't finish after its caller (some magic coroutine thing that messes with the stack maybe?).

Reply:
I see, stack should work fine I think but as you say, it's not performance critical so whatever works

Reply:
I was overthinking this. I changed it to a stack, it's simpler this way.